README.md
# Pypco ReadMe

[![Pypi Version](https://img.shields.io/pypi/v/pypco)](https://pypi.org/project/pypco/) [![Documentation Status](https://readthedocs.org/projects/pypco/badge/?version=latest)](https://pypco.readthedocs.io/en/latest/?badge=latest) [![Build](https://github.com/billdeitrick/pypco/actions/workflows/build.yml/badge.svg)](https://github.com/billdeitrick/pypco/actions/workflows/build.yml?query=branch%3Amaster) [![codecov](https://codecov.io/gh/billdeitrick/pypco/branch/master/graph/badge.svg)](https://codecov.io/gh/billdeitrick/pypco) [![Pypi Python Versions](https://img.shields.io/pypi/pyversions/pypco)](https://pypi.org/project/pypco/)

Stop writing boilerplate code to communicate with the [Planning Center Online](https://planning.center) REST API, and start using pypco! Pypco is a Python wrapper library that supports the full breadth of the PCO REST API. With pypco, you'll spend less time worrying about building and managing HTTP requests and more time building cool things.

[>>> Read the Docs (stable) <<<](https://pypco.readthedocs.io/en/stable/) 

## Features

* *Boilerplate Done for You:* No need to manage HTTP requests, and useful helper functions included (authentication, iteration/pagination, new object templating, etc.)! 
* *Automatic rate limit handling:* If you hit your rate limit, pypco will automatically pause requests and continue once your rate limit period has expired.
* *Automatic Pagination:* Pypco provides automatic pagination/object iteration, allowing you to quickly and easily retrieve large numbers of objects off from the PCO API.
* *Simple Wrapper API:* Using API wrappers can feel like learning *two* new APIs: the REST API itself and the wrapper you're using. Pypco's simple approach is built around the HTTP verbs you already know: GET, POST, PATCH, and DELETE. As a result, after a few minutes with the pypco docs you'll be spending your time in the PCO API docs instead and be on your way to getting things done.
* *Full API Support:* Pypco supports all versions of the PCO v2 REST API, and supports any combination of API versions you might use for each of the PCO apps.

## Examples

```python
import pypco
pco = pypco.PCO("<app_id>", "<app_secret>")

# Print first and last names of everyone in People
for person in pco.iterate('/people/v2/people'):
    print(f'{person["data"]["attributes"]["first_name"]} '\
    f'{person["data"]["attributes"]["last_name"]}')

# Create, save, and print a new person's attribs
payload = pco.template(
  'Person',
  {'first_name': 'John', 'last_name': 'Doe'}
)
new_person = pco.post('/people/v2/people', payload)
print(new_person['data']['attributes'])

# Change our new person's last name and print attribs
payload = pco.template(
  'Person',
  {'last_name': 'Smith'}
)
updated_person = pco.patch(
  f'/people/v2/people/{new_person["data"]["id"]}',
  payload
)
print(updated_person['data']['attributes'])

# Add an email address for our person
payload = pco.template(
  'Email',
  {
    'address': 'john.doe@mailinator.com',
    'location': 'Home'
  }
)
email_object = pco.post(
  f'/people/v2/people/{updated_person["data"]["id"]}/emails',
  payload
)

# Iterate through our person's email addresses
for email in pco.iterate(
  f'/people/v2/people/{updated_person["data"]["id"]}/emails'
):
  print(email['data']['attributes']['address'])
  
```

## Version 1.0

Code written for pypco v0 will not be compatible with the v1 release. Because of changes in the PCO API (primarily the introduction of versioning) and the need for significantly improved performance, v1 is almost a complete rewrite. The result is a much more flexible, performant, and robust API wrapper. Though perhaps a bit less "pythonic", pypco v1.0.x will be much more maintainable going forward. 

If you're relying on pypco v0, you can still find the source code [here](https://github.com/billdeitrick/pypco/tree/v0). Development will not continue on the v0 release, but feel free to submit an issue if you're relying on v0 and need help with a specific problem.

## License

Pypco is licensed under the terms of the MIT License.

CHANGELOG.txt
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.2.0] - 2023-03-03
### Added
- New build pipeline on Github Actions
- Add typing and mypy testing (thanks @07pepa!)
- Add Church Center Org Token authentication (thanks @pastorhudson!)
- Add support and testing for Python versions 3.7 - 3.11

### Changed
- Readability improvements: replace .format() with f strings (thanks @07pepa!)

### Fixed
- Fix 204 response handling from PCO API (thanks @warmach, @pastorhudson!)

## [1.1.0] - 2020-03-10
### Added
- Utilize a session object to enable connection pooling

## [1.0.0] - 2020-01-25
### Changed
- 1.0 is almost a complete rewrite, and is not backwards compatible with v0. There are too many changes to list here, so check out the [docs](https://pypco.readthedocs.io/en/stable/).

## [0.0.2] - 2019-01-01
### Fixed
- Request timeout handling added. HTTP calls that never generate a response will no longer cause applications to hang. If three subsequent HTTP calls time out, an exception will be raised.
- Order and per_page parameters not being passed when using rel.list()

## [0.0.1] - 2018-11-26
### Changed
- Now using the User-Agent "pypco" for all HTTP calls to the PCO API 

### Fixed
- Fix missing imports preventing Resources and Webhooks APIs from working

## [0.0.0] - 2018-11-24
### Added
- Initial alpha release

pypco/pco.py
"""The primary module for pypco containing main wrapper logic."""

import time
import logging
import re

from typing import Any, Iterator, Optional
import requests

from .auth_config import PCOAuthConfig
from .exceptions import PCORequestTimeoutException, \
    PCORequestException, PCOUnexpectedRequestException


class PCO:  # pylint: disable=too-many-instance-attributes
    """The entry point to the PCO API.

    Note:
        You must specify either an application ID and a secret or an oauth token.
        If you specify an invalid combination of these arguments, an exception will be
        raised when you attempt to make API calls.

    Args:
        application_id (str): The application_id; secret must also be specified.
        secret (str): The secret for your app; application_id must also be specified.
        token (str): OAUTH token for your app; application_id and secret must not be specified.
        api_base (str): The base URL against which REST calls will be made.
            Default: https://api.planningcenteronline.com
        timeout (int): How long to wait (seconds) for requests to timeout. Default 60.
        upload_url (str): The URL to which files will be uploaded.
            Default: https://upload.planningcenteronline.com/v2/files
        upload_timeout (int): How long to wait (seconds) for uploads to timeout. Default 300.
        timeout_retries (int): How many times to retry requests that have timed out. Default 3.
    """

    def __init__(  # pylint: disable=too-many-arguments
            self,
            application_id: Optional[str] = None,  # pylint: disable=unsubscriptable-object
            secret: Optional[str] = None,  # pylint: disable=unsubscriptable-object
            token: Optional[str] = None,  # pylint: disable=unsubscriptable-object
            cc_name: Optional[str] = None,  # pylint: disable=unsubscriptable-object
            api_base: str = 'https://api.planningcenteronline.com',
            timeout: int = 60,
            upload_url: str = 'https://upload.planningcenteronline.com/v2/files',
            upload_timeout: int = 300,
            timeout_retries: int = 3,
    ):

        self._log = logging.getLogger(__name__)

        self._auth_config = PCOAuthConfig(application_id, secret, token, cc_name)
        self._auth_header = self._auth_config.auth_header

        self.api_base = api_base
        self.timeout = timeout

        self.upload_url = upload_url
        self.upload_timeout = upload_timeout

        self.timeout_retries = timeout_retries

        self.session = requests.Session()

        self._log.debug("Pypco has been initialized!")

    def _do_request(
            self,
            method: str,
            url: str,
            payload: Optional[Any] = None,  # pylint: disable=unsubscriptable-object
            upload: Optional[str] = None,  # pylint: disable=unsubscriptable-object
            **params
    ) -> requests.Response:
        """Builds, executes, and performs a single request against the PCO API.

        Executed request could be one of the standard HTTP verbs or a file upload.

        Args:
            method (str): The HTTP method to use for this request.
            url (str): The URL against which this request will be executed.
            payload (obj): A json-serializable Python object to be sent as the post/put payload.
            upload(str): The path to a file to upload.
            params (obj): A dictionary or list of tuples or bytes to send in the query string.

        Returns:
            requests.Response: The response to this request.
        """

        # Standard header
        headers = {
            'User-Agent': 'pypco',
            'Authorization': self._auth_header,
        }

        # Standard params
        request_params = {
            'headers': headers,
            'params': params,
            'json': payload,
            'timeout': self.upload_timeout if upload else self.timeout
        }

        # Add files param if upload specified
        if upload:
            upload_fh = open(upload, 'rb')
            request_params['files'] = {'file': upload_fh}

        self._log.debug(
            "Executing %s request to '%s' with args %s",
            method,
            url,
            {param: value for (param, value) in request_params.items() if param != 'headers'}
        )

        # The moment we've been waiting for...execute the request
        try:
            response = self.session.request(
                method,
                url,
                **request_params # type: ignore[arg-type]
            )
        finally:
            if upload:
                upload_fh.close()

        return response

    def _do_timeout_managed_request(
            self,
            method: str,
            url: str,
            payload: Optional[Any] = None,  # pylint: disable=unsubscriptable-object
            upload: Optional[str] = None,  # pylint: disable=unsubscriptable-object
            **params
        ) -> requests.Response:
        """Performs a single request against the PCO API with automatic retried in case of timeout.

        Executed request could be one of the standard HTTP verbs or a file upload.

        Args:
            method (str): The HTTP method to use for this request.
            url (str): The URL against which this request will be executed.
            payload (obj): A json-serializable Python object to be sent as the post/put payload.
            upload(str): The path to a file to upload.
            params (obj): A dictionary or list of tuples or bytes to send in the query string.

        Raises:
            PCORequestTimeoutException: The request to PCO timed out the maximum number of times.

        Returns:
            requests.Response: The response to this request.
        """

        timeout_count = 0

        while True:
            try:
                return self._do_request(method, url, payload, upload, **params)

            except requests.exceptions.Timeout as exc:
                timeout_count += 1

                self._log.debug("The request to \"%s\" timed out after %d tries.",
                                url, timeout_count)

                if timeout_count == self.timeout_retries:
                    self._log.debug("Maximum retries (%d) hit. Will raise exception.",
                                    self.timeout_retries)

                    raise PCORequestTimeoutException(
                        f"The request to \"{url}\" timed out after {timeout_count} tries.") from exc

                continue

    def _do_ratelimit_managed_request(
            self,
            method: str,
            url: str,
            payload: Optional[Any] = None,  # pylint: disable=unsubscriptable-object
            upload: Optional[str] = None,  # pylint: disable=unsubscriptable-object
            **params
        ) -> requests.Response:
        """Performs a single request against the PCO API with automatic rate limit handling.

        Executed request could be one of the standard HTTP verbs or a file upload.

        Args:
            method (str): The HTTP method to use for this request.
            url (str): The URL against which this request will be executed.
            payload (obj): A json-serializable Python object to be sent as the post/put payload.
            upload(str): The path to a file to upload.
            params (obj): A dictionary or list of tuples or bytes to send in the query string.

        Raises:
            PCORequestTimeoutException: The request to PCO timed out the maximum number of times.

        Returns:
            requests.Response: The response to this request.
        """

        while True:

            response = self._do_timeout_managed_request(method, url, payload, upload, **params)

            if response.status_code == 429:
                self._log.debug("Received rate limit response. Will try again after %d sec(s).",
                                int(response.headers['Retry-After']))

                time.sleep(int(response.headers['Retry-After']))
                continue

            return response

    def _do_url_managed_request(
            self,
            method: str,
            url: str,
            payload: Optional[Any] = None,  # pylint: disable=unsubscriptable-object
            upload: Optional[str] = None,  # pylint: disable=unsubscriptable-object
            **params
        ) -> requests.Response:
        """Performs a single request against the PCO API, automatically cleaning up the URL.

        Executed request could be one of the standard HTTP verbs or a file upload.

        Args:
            method (str): The HTTP method to use for this request.
            url (str): The URL against which this request will be executed.
            payload (obj): A json-serializable Python object to be sent as the post/put payload.
            upload(str): The path to a file to upload.
            params (obj): A dictionary or list of tuples or bytes to send in the query string.

        Raises:
            PCORequestTimeoutException: The request to PCO timed out the maximum number of times.

        Returns:
            requests.Response: The response to this request.
        """

        self._log.debug("URL cleaning input: \"%s\"", url)

        if not upload:
            url = url if url.startswith(self.api_base) else f'{self.api_base}{url}'
            url = re.subn(r'(?<!:)[/]{2,}', '/', url)[0]

        self._log.debug("URL cleaning output: \"%s\"", url)

        return self._do_ratelimit_managed_request(method, url, payload, upload, **params)

    def request_response(
            self,
            method: str,
            url: str,
            payload: Optional[Any] = None,  # pylint: disable=unsubscriptable-object
            upload: Optional[str] = None,  # pylint: disable=unsubscriptable-object
            **params
        ) -> requests.Response:
        """A generic entry point for making a managed request against PCO.

        This function will return a Requests response object, allowing access to
        all request data and metadata. Executed request could be one of the standard
        HTTP verbs or a file upload. If you're just looking for your data (json), use
        the request_json() function or get(), post(), etc.

        Args:
            method (str): The HTTP method to use for this request.
            url (str): The URL against which this request will be executed.
            payload (obj): A json-serializable Python object to be sent as the post/put payload.
            upload(str): The path to a file to upload.
            params (obj): A dictionary or list of tuples or bytes to send in the query string.

        Raises:
            PCORequestTimeoutException: The request to PCO timed out the maximum number of times.
            PCOUnexpectedRequestException: An unexpected error occurred when making your request.
            PCORequestException: The response from the PCO API indicated an error with your request.

        Returns:
            requests.Response: The response to this request.
        """

        try:
            response = self._do_url_managed_request(method, url, payload, upload, **params)
        except Exception as err:
            self._log.debug("Request resulted in unexpected error: \"%s\"", str(err))
            raise PCOUnexpectedRequestException(str(err)) from err

        try:
            response.raise_for_status()
        except requests.HTTPError as err:
            self._log.debug("Request resulted in API error: \"%s\"", str(err))
            raise PCORequestException(
                response.status_code,
                str(err),
                response_body=response.text
            ) from err

        return response

    def request_json(
            self,
            method: str,
            url: str,
            payload: Optional[Any] = None,  # pylint: disable=unsubscriptable-object
            upload: Optional[str] = None,  # pylint: disable=unsubscriptable-object
            **params: str
    ) -> Optional[dict]:  # pylint: disable=unsubscriptable-object
        """A generic entry point for making a managed request against PCO.

        This function will return the payload from the PCO response (a dict).

        Args:
            method (str): The HTTP method to use for this request.
            url (str): The URL against which this request will be executed.
            payload (obj): A json-serializable Python object to be sent as the post/put payload.
            upload(str): The path to a file to upload.
            params (obj): A dictionary or list of tuples or bytes to send in the query string.

        Raises:
            PCORequestTimeoutException: The request to PCO timed out the maximum number of times.
            PCOUnexpectedRequestException: An unexpected error occurred when making your request.
            PCORequestException: The response from the PCO API indicated an error with your request.

        Returns:
            dict: The payload from the response to this request.
        """

        response = self.request_response(method, url, payload, upload, **params)
        if response.status_code == 204:
            return_value = None
        else:
            return_value = response.json()

        return return_value

    def get(self, url: str, **params) -> Optional[dict]:  # pylint: disable=unsubscriptable-object
        """Perform a GET request against the PCO API.

        Performs a fully managed GET request (handles ratelimiting, timeouts, etc.).

        Args:
            url (str): The URL against which to perform the request. Can include
                what's been set as api_base, which will be ignored if this value is also
                present in your URL.
            params: Any named arguments will be passed as query parameters.

        Raises:
            PCORequestTimeoutException: The request to PCO timed out the maximum number of times.
            PCOUnexpectedRequestException: An unexpected error occurred when making your request.
            PCORequestException: The response from the PCO API indicated an error with your request.

        Returns:
            dict: The payload returned by the API for this request.
        """

        return self.request_json('GET', url, **params)

    def post(
            self,
            url: str,
            payload: Optional[dict] = None,  # pylint: disable=unsubscriptable-object
            **params: str
        ) -> Optional[dict]:  # pylint: disable=unsubscriptable-object
        """Perform a POST request against the PCO API.

        Performs a fully managed POST request (handles ratelimiting, timeouts, etc.).

        Args:
            url (str): The URL against which to perform the request. Can include
                what's been set as api_base, which will be ignored if this value is also
                present in your URL.
            payload (dict): The payload for the POST request. Must be serializable to JSON!
            params: Any named arguments will be passed as query parameters. Values must
                be of type str!

        Raises:
            PCORequestTimeoutException: The request to PCO timed out the maximum number of times.
            PCOUnexpectedRequestException: An unexpected error occurred when making your request.
            PCORequestException: The response from the PCO API indicated an error with your request.

        Returns:
            dict: The payload returned by the API for this request.
        """

        return self.request_json('POST', url, payload, **params)

    def patch(
            self,
            url: str,
            payload: Optional[dict] = None,  # pylint: disable=unsubscriptable-object
            **params: str
        ) -> Optional[dict]:  # pylint: disable=unsubscriptable-object
        """Perform a PATCH request against the PCO API.

        Performs a fully managed PATCH request (handles ratelimiting, timeouts, etc.).

        Args:
            url (str): The URL against which to perform the request. Can include
                what's been set as api_base, which will be ignored if this value is also
                present in your URL.
            payload (dict): The payload for the PUT request. Must be serializable to JSON!
            params: Any named arguments will be passed as query parameters. Values must
                be of type str!

        Raises:
            PCORequestTimeoutException: The request to PCO timed out the maximum number of times.
            PCOUnexpectedRequestException: An unexpected error occurred when making your request.
            PCORequestException: The response from the PCO API indicated an error with your request.

        Returns:
            dict: The payload returned by the API for this request.
        """

        return self.request_json('PATCH', url, payload, **params)

    def delete(self, url: str, **params: str) -> requests.Response:
        """Perform a DELETE request against the PCO API.

        Performs a fully managed DELETE request (handles ratelimiting, timeouts, etc.).

        Args:
            url (str): The URL against which to perform the request. Can include
                what's been set as api_base, which will be ignored if this value is also
                present in your URL.
            params: Any named arguments will be passed as query parameters. Values must
                be of type str!

        Raises:
            PCORequestTimeoutException: The request to PCO timed out the maximum number of times.
            PCOUnexpectedRequestException: An unexpected error occurred when making your request.
            PCORequestException: The response from the PCO API indicated an error with your request.

        Returns:
            requests.Response: The response object returned by the API for this request.
            A successful delete request will return a response with an empty payload,
            so we return the response object here instead.
        """

        return self.request_response('DELETE', url, **params)

    def iterate(self, url: str, offset: int = 0, per_page: int = 25, **params: str) -> Iterator[dict]:  # pylint: disable=too-many-branches
        """Iterate a list of objects in a response, handling pagination.

        Basically, this function wraps get in a generator function designed for
        processing requests that will return multiple objects. Pagination is
        transparently handled.

        Objects specified as includes will be injected into their associated
        object and returned.

        Args:
            url (str): The URL against which to perform the request. Can include
                what's been set as api_base, which will be ignored if this value is also
                present in your URL.
            offset (int): The offset at which to start. Usually going to be 0 (the default).
            per_page (int): The number of results that should be requested in a single page.
                Valid values are 1 - 100, defaults to the PCO default of 25.
            params: Any additional named arguments will be passed as query parameters. Values must
                be of type str!

        Raises:
            PCORequestTimeoutException: The request to PCO timed out the maximum number of times.
            PCOUnexpectedRequestException: An unexpected error occurred when making your request.
            PCORequestException: The response from the PCO API indicated an error with your request.

        Yields:
            dict: Each object returned by the API for this request. Returns "data",
            "included", and "meta" nodes for each response. Note that data is processed somewhat
            before being returned from the API. Namely, includes are injected into the object(s)
            with which they are associated. This makes it easier to process includes associated with
            specific objects since they are accessible directly from each returned object.
        """

        while True:  # pylint: disable=too-many-nested-blocks

            response = self.get(url, offset=offset, per_page=per_page, **params)

            if response is None:
                return

            for cur in response['data']:
                record = {
                    'data': cur,
                    'included': [],
                    'meta': {}
                }

                if 'can_include' in response['meta']:
                    record['meta']['can_include'] = response['meta']['can_include']

                if 'parent' in response['meta']:
                    record['meta']['parent'] = response['meta']['parent']

                if 'relationships' in cur:
                    for key in cur['relationships']:
                        relationships = cur['relationships'][key]['data']

                        if relationships is not None:
                            if isinstance(relationships, dict):
                                for include in response['included']:
                                    if include['type'] == relationships['type'] and \
                                            include['id'] == relationships['id']:
                                        record['included'].append(include)

                            elif isinstance(relationships, list):
                                for relationship in relationships:
                                    for include in response['included']:
                                        if include['type'] == relationship['type'] and \
                                                include['id'] == relationship['id']:
                                            record['included'].append(include)

                yield record

            offset += per_page

            if 'next' not in response['links']:
                break

    def upload(self, file_path: str, **params) -> Optional[dict]:  # pylint: disable=unsubscriptable-object
        """Upload the file at the specified path to PCO.

        Args:
            file_path (str): The path to the file to be uploaded to PCO.
            params: Any named arguments will be passed as query parameters. Values must
                be of type str!

        Raises:
            PCORequestTimeoutException: The request to PCO timed out the maximum number of times.
            PCOUnexpectedRequestException: An unexpected error occurred when making your request.
            PCORequestException: The response from the PCO API indicated an error with your request.

        Returns:
            dict: The PCO response from the file upload.
        """

        return self.request_json('POST', self.upload_url, upload=file_path, **params)

    def __del__(self):
        """Close the requests session when the PCO object goes out of scope."""

        self.session.close()

    @staticmethod
    def template(
            object_type: str,
            attributes: Optional[dict] = None  # pylint: disable=unsubscriptable-object
    ) -> dict:
        """Get template JSON for creating a new object.

        Args:
            object_type (str): The type of object to be created.
            attributes (dict): The new objects attributes. Defaults to empty.

        Returns:
            dict: A template from which to set the new object's attributes.
        """

        return {
            'data': {
                'type': object_type,
                'attributes': {} if attributes is None else attributes
            }
        }
